crates/runtime/src/lib.rs

```diff
@@ -610,15 +658,72 @@ impl Runtime {
     }
 
     // Closes and deallocates all resources (including the static registries)
-    pub async fn shutdown(self) {
+    pub async fn shutdown(&self) {
+        if self.status.is_shutdown() {
+            return;
+        }
+
+        tracing::info!("Shutting down runtime...");
         self.status.mark_shutdown();
+
+        // shutdown all running components except the HTTP and Metrics servers
+        let mut running_components = self.server_components.write().await;
+
+        // HTTP and METRICS servers must be shutdown last
+        let (to_shutdown, to_shutdown_last) = running_components
+            .drain()
+            .partition::<Vec<_>, _>(|(name, _)| *name != HTTP_SERVER && *name != METRICS_SERVER);
+
+        let shutdown_futures: Vec<_> = to_shutdown
+            .into_iter()
+            .map(|(name, handle)| {
+                tracing::debug!("Shutting down {name}");
+                handle.shutdown(SERVER_COMPONENT_SHUTDOWN_TIMEOUT)
+            })
+            .collect();
+
+        join_all(shutdown_futures).await;
+
         // Clean up DataFusion first as there could be datasets loading and accessing registries below.
         self.df.shutdown().await;
         dataconnector::unregister_all().await;
         catalogconnector::unregister_all().await;
         dataaccelerator::unregister_all().await;
         tools::factory::unregister_all_factories().await;
         document_parse::unregister_all().await;
```

Comment:
Not for this PR, but I think we should move to the model where all static registries are tied to the runtime object instead of being static. This code essentially prevents us from having multiple runtimes defined, which is useful for testing.